### DISCUSSION TITLE
Local History for Files

### DISCUSSION THREAD

**Discussion by x80486** [zed-industries/zed#24004] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

I would like to suggest an enhancement that I'm sure could benefit many users: the addition of a _local history_ feature for files, similar to the one available in IntelliJ IDEA.

IntelliJ IDEA provides a local history feature that automatically records changes to files and sub-directories **independently** of version control systems. This comes pretty handy to view a timeline of changes for each file, and compare (or restore) different (or previous) versions of a given file or sub-directory.

### If applicable, add mockups / screenshots to help present your vision of the feature

![image](https://github.com/user-attachments/assets/117f8bd5-98e5-4caa-af4f-2c1fff97193d)

**Comment by gevgeny** [comment]
I would say lack of local history is the last reason why I cannot fully change my IDEA to Zed. I could reconcile with lack of other features, but this one is too useful. 

Here is one more screenshot from IDEA which shows what we need: 

<img width="1238" alt="image" src="https://github.com/user-attachments/assets/c02e750f-1d27-4e5b-ad54-e90f1ad7d91b">

**Comment by prasanna4742** [comment]
I'm in love with Zed, we really need this feature, I was struggling with mac keys and nearly lost my 4 hours of work because there's no local history (I know it's on me, but would really appreciate local history)

**Comment by devnulled** [comment]
+1, I'm really digging Zed and this is the only major feature I can think of at the moment that it needs as sort of a baseline type feature.  I think it's necessary as a base functionality of an editor/IDE rather than being a "nice to have".  

Right now I'm having to do other hacky things in case I want to go back to previous versions of files.  This feature in the JetBrains suite is a lifesaver for so many reasons.. like not worrying about leaving auto-save on, not worried about the UI crashing and losing your work, etc.

Thanks for all of the great work!  Just adding a +1 in hopes of it helping it move up in priority.

**Comment by rezzubs** [comment]
Not sure how the history works in JetBrains but it would be cool if the history could be a tree like in vim. This way you can never lose existing states when making edits at a previous point.

**Reply by zedmakesense** [reply]
I think vim tree would be the best implementation here. like the one which shows the branches too. (there was a plugin in nvim for that too (for branches))

**Comment by trymeouteh** [comment]
VSCode has a local history feature which I cannot live without and would like to see in Zed.

**Comment by MaartenPI** [comment]
Same here, needed as well

**Comment by simon-something** [comment]
I've landed here following an "accidental" `git restore .` (opening zed opens all previous windows, didn't check which project was open and..boom, I restored the wrong one). Any other IDE would have saved me - please prioritise this, it *really* is useful

**Comment by JosephTLyons** [comment]
I think once we have git diff components in Zed, and once we put in the work for CRDB, we'd use CRDB to show local file changes within the git diff view. Seems like something we need to do, especially for people who have lost code due to some user error, or some Zed bug.

**Reply by negentropi** [reply]
today is jul 22 2025. i think we close

**Comment by mattijsf** [comment]
This would be very nice. Maybe it can be done through an extension like https://marketplace.visualstudio.com/items?itemName=xyz.local-history

**Comment by mattijsf** [comment]
I switched to Zed but then I'd like to revert something that wasn't yet stashed / committed and then I thought... damn I had this in vscode 😅

**Comment by ikapta** [comment]
Sadly, after spending the morning writing code and successfully debugging it in one go, I happily stepped away to pour a glass of water—only to return and find that one of my files had reverted to its initial version. I’ve recommended Zed to so many colleagues because I genuinely love it, Thank you all for your amazing work—I really appreciate it!

**Comment by rdvo** [comment]
yep zed is great i tried to switch from cursor to use the new agentic stuff but the lack of a historu/"timeline" view to go thru undo/edits along with the fact when u go back in an agent chat and edit the text it does not revert the changes back to that state

would love fo upgrade to paid if they fixed/adde this!

**Comment by vitorfdl** [comment]
I was having fun with Zed until I noticed it was missing this feature. Local file history has saved my ass so many times that I don't have the confidence to work without it.

Also, Git is not a substitute. I think both features complement each other.

**Comment by negentropi** [comment]
yeah, thats why i didnot switch to zed. search zed's extension also none. 

zed, like this we cant together.

**Comment by aaron-18-112** [comment]
+1, This would be a great feature

**Comment by hrprice** [comment]
+1, this is one of the main things keeping me from switching to zed. 

If any zed devs happen to read this, is this something that's being worked on, or would something like this be better off as an extension?

**Comment by warrenjokinen** [comment]
I would like to see a change from the ancient **"Blame" to something like "kudos", "kiitos", or something.**

**Comment by hennessyevan** [comment]
Tried zed again after about a year and was excited to see all my most used features from vscode now in zed! But... no file history is a big deal since I like to be able to scan through older versions of a file(s) especially when experimenting.

**Comment by zavierrodriguez** [comment]
Would love to see this feature as well.  This and snippets (less so) are features I really need to make it my daily driver

**Comment by dmtrkzntsv** [comment]
Zed is fantastic, but the only feature missing is the local history plugin.  It’s especially useful when working with LLMs

**Comment by MBestKing** [comment]
Is it still in dev or its out there? i really need this feature its great, so great that you almost can't live with out it.

**Reply by ovidius72** [reply]
I think it's available in the Preview release

**Reply by Helmchen-Lord** [reply]
They have only implemented a history for git commits (which was already included in version 0.216.0), but unfortunately not as it is in IntelliJ IDEA (unless I missed something).

**Comment by Kyaee** [comment]
+1, hope we can add this tracking feature

**Comment by ivan-dalmet** [comment]
Any idea if an extension or a built in solution is planned or available for this?
This is maybe the thing that will make me switch back 😟

**Reply by o-oOvOo-o** [reply]
I think I already made it. However, there are a bit of  changes, which makes it less suitable for contribution. I'm considering I should have a hard fork instead and wait their contribution temself.
<img width="344" height="377" alt="image" src="https://github.com/user-attachments/assets/9a62747b-e922-48ed-b97d-c41ad1d01051" />
![Uploading image.png…]()

**Reply by o-oOvOo-o** [reply]
Perhaps the history logic need separated,  but it interface would be more appropriate to combine to show in the git panel like vscode. maybe the dev mean is : history should be together with git on the interface.

**Reply by singalen** [reply]
Hard fork won't get you users, but I'd love to see it integrated with git views in Zed. `Mnemosyne` project down in the comments looks very relevant too.

**Comment by htahboub** [comment]
Hi @JosephTLyons, is this still planned?

**Comment by maximxlss** [comment]
+1, also pretty much the only thing currently keeping me from switching. This is a really important feature to me, I'm using it quite often.

**Comment by alessandrobrunoh** [comment]
+1 i hope this get pr fast, maybe its me but i always use it a lot.

(for now i had to make my own CLI that track every changes)
https://github.com/alessandrobrunoh/Mnemosyne

**Comment by jonasbaldwin** [comment]
+1

**Comment by swannysec** [comment]
Hey all!  Please consider using the built-in upvote or reactions on the main discussion comment instead of making `+1` replies that do not add to the conversation or contain additional context.  Simple `+1` comments don't meaningfully add signal and increase noise for the author and other folks subscribed to the discussion.

Thanks!

**Comment by magicdude4eva** [comment]
I switched to Zed only to learn that local history is not a feature. True pity, back to vscode it is :-(

**Comment by ejohb** [comment]
PyCharm's Local History has saved my blue, blue bacon more times than I can possibly count. Please consider implementing.

**Comment by borges-filipe** [comment]
This is the only feature stopping me from switching over to Zed for good. Any performance and quality of life improvements are nothing compared to me being able to retrieve a possibly important file I accidentally deleted.

**Comment by ejohb** [comment]
Even a local-only branch to the existing repo, that records every save as a commit, would be 80% of the way there. Still not trivial to implement, I'm sure, but Git already does the heavy lifting, is my point, and there's no immediate need for a separate, Zed-specific history store.

**Comment by borges-filipe** [comment]
I believe this issue has not been addressed by the Zed team because they are hoping to address this with their [new version control system (DeltaDB)](https://zed.dev/blog/introducing-deltadb). I wonder if this solution will have any benefits for non-ai workflows.

**Comment by trymeouteh** [comment]
Would like a non-git local history feature which stores changes in a hidden directory on the device like in VSCode.
